0000-open-options.md

```diff
@@ -0,0 +1,636 @@
+undefined
+(see [some](http://www.monkey.org/openbsd/archive/tech/0009/msg00299.html)
+[comments](http://www.monkey.org/openbsd/archive/tech/0009/msg00304.html) on the
+OpenBSD mailing list). The behaviour on Windows is inconsistent and depends on
+whether `.create(true)` is set.
+
+To give guarantees about cross-platform (and sane) behaviour, Rust should allow
+only the following combinations of access modes and creations modes:
+
+creation mode           | read  | write | read-write | append | read-append |
+:-----------------------|:-----:|:-----:|:----------:|:------:|:-----------:|
+not set (open existing) |   X   |   X   |     X      |   X    |      X      |
+create                  |       |   X   |     X      |   X    |      X      |
+truncate                |       |   X   |     X      |        |             |
+create and truncate     |       |   X   |     X      |        |             |
+create_new              |       |   X   |     X      |   X    |      X      |
+
+It is possible to bypass these restrictions by using system-specific options (as
+in this case you already have to take care of cross-platform support yourself).
+On Unix this is done by setting the creation mode using `.custom_flags()` with
+`O_CREAT`, `O_TRUNC` and/or `O_EXCL`. On Windows this can be done by manually
+specifying `.access_mode()` (see above).
+
+
+## Sharing / locking
+On Unix it is possible for multiple processes to read and write to the same file
+at the same time.
+
+When you open a file on Windows, the system by default denies other processes to
+read or write to the file, or delete it. By setting the sharing mode, it is
+possible to allow other processes read, write and/or delete access. For
+cross-platform consistency, Rust imitates Unix by setting all sharing flags.
+
+Unix has no equivalent to the kind of file locking that Windows has. It has two
+types of advisory locking, POSIX and BSD-style. Advisory means any process that
+does not use locking itself can happily ignore the locking af another process.
+As if that is not bad enough, they both have
+[problems](http://0pointer.de/blog/projects/locking.html) that make them close
+to unusable for modern multi-threaded programs. Linux may in some very rare
+cases support mandatory file locking, but it is just as broken as advisory.
+
+For Rust, the sharing mode can be set with a Windows-specific option. Given the
+problems above, I don't expect there to ever be a cross-platform option for file
+locking.
+
+
+### Windows-specific: Share mode
+`.share_mode(FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE)`
+
+It is possible to set the individual share permissions with `.share_mode()`.
+
+The current philosophy of this function is that others should have no rights,
+unless explicitly granted. I think a better fit for Rust would be to give all
+others all rights, unless explicitly denied, e.g.:
+`.share_mode(DENY_READ | DENY_WRITE | DENY_DELETE)`.
+
+
+## Caching behaviour
```

Comment:
Could you elaborate a bit on the inclusion of these new options? I would personally expect this to fall under the category of "out of scope for now" in the sense that this should all in theory be possible to implement externally (e.g. using the as-raw-os-handle style traits).